pkg/storage/chunkfs/chunk.go

```diff
@@ -0,0 +1,422 @@
+	oldMBuf, err := c.mmf.Buffer(mOffset, mSize)
+	if err != nil {
+		c.logger.Errorf("could not map meta-buffer with offset %d for size=%d: %v", mOffset, mSize, err)
+		return fmt.Errorf("could not Grow Chunk: %w", fmt.Errorf("could not map meta-buffer with offset %d for size=%d: %w", mOffset, mSize, errors.ErrInternal))
+	}
+	newMBuf, err := c.mmf.Buffer(c.mmf.Size()-int64(mSize), mSize)
+	if err != nil {
+		c.logger.Errorf("could not map meta-buffer with offset %d for size=%d: %v", mOffset, mSize, err)
+		return fmt.Errorf("could not Grow Chunk: %w", fmt.Errorf("could not map meta-buffer with offset %d for size=%d: %w", mOffset, mSize, errors.ErrInternal))
+	}
+	copy(newMBuf, oldMBuf)
+
+	return nil
+}
+
+// OpenChunkReader opens new read operation. The function returns ChunkReader, which may be used for reading the chunk
+// records. The ChunkReader must be closed. The AppendRecords and Close() operations will be blocked until ALL
+// ChunkReaders are closed. So the ChunkReader should be requested for a short period of time and be closed as soon as
+// possible
+func (c *Chunk) OpenChunkReader(descending bool) (*ChunkReader, error) {
+	c.lock.RLock()
+	if c.mmf == nil {
+		// chunk is closed
+		c.lock.RUnlock()
+		return nil, fmt.Errorf("the chunk %s is closed: %w ", c.fn, errors.ErrClosed)
+	}
+	mb, err := c.getMetaBuf(c.total-1, c.total)
+	if err != nil {
+		return nil, err
+	}
+
+	cr := new(ChunkReader)
+	cr.c = c
+	cr.inc = 1
+	cr.mb = mb
+
+	if descending {
+		cr.inc = -1
+		cr.idx = cr.c.total - 1
+	}
+	return cr, nil
+}
+
+func (c *Chunk) available() int64 {
+	return c.mmf.Size() - int64(c.freeOffset+c.total*cMetaRecordSize)
+}
+
+func payloadSize(recs []*solaris.Record) int64 {
+	size := int64(0)
+	for _, r := range recs {
+		size += int64(len(r.Payload))
+	}
+	return size
+}
+
+func (cr *ChunkReader) HasNext() bool {
+	return cr.idx < cr.c.total && cr.idx > -1
+}
+
+func (cr *ChunkReader) Next() (UnsafeRecord, bool) {
+	if cr.HasNext() {
+		mr := cr.mb.get(cr.idx)
+		buf, err := cr.c.mmf.Buffer(int64(mr.offset), int(mr.size))
+		if err != nil {
+			cr.c.logger.Errorf("could not read payload for offset=%d for len=%d: %v", mr.offset, mr.size, err)
+			panic(err)
+		}
+		res := UnsafeRecord{ID: mr.ID, UnsafePayload: buf}
+		cr.idx += cr.inc
+		return res, true
+	}
+	return UnsafeRecord{}, false
+}
+
+// Close implements io.Closer
+func (cr *ChunkReader) Close() error {
+	cr.c.lock.RUnlock()
+	cr.c = nil
+	return nil
+}
+
+// SetStartID moves the iterator offset to the position startID. The function returns the number of records
+// which will be available for read after the call taking into account the direction of the iterator.
+func (cr *ChunkReader) SetStartID(startID ulid.ULID) int {
+	res := 0
+	if cr.inc == -1 {
+		cr.idx = sort.Search(cr.c.total, func(i int) bool {
+			return cr.mb.get(i).ID.Compare(startID) > 0
+		})
+		cr.idx--
+		res = cr.idx + 1
+	} else {
+		cr.idx = sort.Search(cr.c.total, func(i int) bool {
```

Comment:
q: Is it ok that `cr.idx` can be actually == `cr.total`? I guess so, since we do `HasNext()` before using `cr.idx`, just confirming

Reply:
`cr.idx` will be `cr.total` if no such element found. This case the iterator will be empty and no iteration happens. In case of the descending order the situation is reversed and we have to walk through the whole set of records.